File: CS/WebLayoutSolution.Module.Web/WebLayoutManagerEx.cs
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Layout;
using DevExpress.ExpressApp.Model;
using DevExpress.ExpressApp.Web.Layout;
using DevExpress.ExpressApp.Web.SystemModule;
using DevExpress.Utils;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;

//#pragma warning disable 109
namespace WebLayoutSolution.Module.Web {

    public enum TextAlignModeGroup {
        UseParentOptions = 0,
        AlignLocal = 1,
        AutoSize = 2,
        //CustomSize = 3,
        AlignWithChildren = 4
    }
    public enum TextAlignModeItem {
        UseParentOptions = 0,
        //AlignLocal = 1,
        AutoSize = 2,
        //CustomSize = 3,
        //AlignWithChildren = 4
    }
    public interface IModelWebLayoutGroup {
        [Category("Behavior"), DefaultValue(TextAlignModeGroup.UseParentOptions)]
        TextAlignModeGroup TextAlignMode { get; set; }
    }
    public interface IModelWebLayoutItem {
        [Category("Behavior"), DefaultValue(TextAlignModeItem.UseParentOptions)]
        TextAlignModeItem TextAlignMode { get; set; }
    }
    public class WebLayoutManagerEx : WebLayoutManager {
        public WebLayoutManagerEx(bool simple, bool delayedItemsInitialization) : base(simple, delayedItemsInitialization) { }
        protected override LayoutItemTemplateContainer LayoutItem(ViewItemsCollection viewItems, IModelLayoutViewItem layoutItemModel) {
            LayoutItemTemplateContainer templateContainer = new LayoutItemTemplateContainer(this, viewItems, layoutItemModel);
            templateContainer.Template = LayoutItemTemplate;
            templateContainer.ID = WebIdHelper.GetCorrectedLayoutItemId(layoutItemModel);
            ViewItem viewItem = FindViewItem(viewItems, layoutItemModel);
            templateContainer.ViewItem = viewItem;
            if (viewItem != null) {
                MarkRequiredFieldCaptionEventArgs args = new MarkRequiredFieldCaptionEventArgs(viewItem, false);
                OnMarkRequiredFieldCaption(args);
                templateContainer.Caption = BuildItemCaption(viewItem, args.NeedMarkRequiredField, args.RequiredFieldMark);
            }
            templateContainer.ShowCaption = GetIsLayoutItemCaptionVisible(layoutItemModel, viewItem);
            templateContainer.CaptionWidth = CalculateCaptionWidth(viewItem, viewItems, layoutItemModel);
            templateContainer.CaptionLocation = GetCaptionLocation(layoutItemModel);
            templateContainer.CaptionHorizontalAlignment = GetCaptionHorizontalAlignment(layoutItemModel);
            templateContainer.CaptionVerticalAlignment = GetCaptionVerticalAlignment(layoutItemModel);
            OnLayoutItemCreatedNew(templateContainer, layoutItemModel, viewItem);
            return templateContainer;
        }
        private void OnLayoutItemCreatedNew(LayoutItemTemplateContainerBase templateContainer, IModelViewLayoutElement layoutItemModel, ViewItem viewItem) {
            if (!DelayedItemsInitialization) {
                templateContainer.Instantiate();
            }
            OnLayoutItemCreated(new ItemCreatedEventArgs(layoutItemModel, viewItem, templateContainer));
            OnCustomizeAppearance(new CustomizeAppearanceEventArgs(layoutItemModel.Id, new WebLayoutItemAppearanceAdapter(templateContainer), null));
        }
        private DevExpress.Utils.Locations GetCaptionLocation(IModelLayoutViewItem layoutItemModel) {
            DevExpress.Utils.Locations captionLocation = layoutItemModel.CaptionLocation;
            return Equals(captionLocation, DevExpress.Utils.Locations.Default) ? DefaultLayoutItemCaptionLocation : captionLocation;
        }
        private DevExpress.Utils.HorzAlignment GetCaptionHorizontalAlignment(IModelLayoutViewItem layoutItemModel) {
            DevExpress.Utils.HorzAlignment captionHorizontalAlignment = layoutItemModel.CaptionHorizontalAlignment;
            return Equals(captionHorizontalAlignment, DevExpress.Utils.HorzAlignment.Default) ? DefaultLayoutItemCaptionHorizontalAlignment : captionHorizontalAlignment;
        }
        private DevExpress.Utils.VertAlignment GetCaptionVerticalAlignment(IModelLayoutViewItem layoutItemModel) {
            DevExpress.Utils.VertAlignment captionVerticalAlignment = layoutItemModel.CaptionVerticalAlignment;
            return Equals(captionVerticalAlignment, DevExpress.Utils.VertAlignment.Default) ? DefaultLayoutItemCaptionVerticalAlignment : captionVerticalAlignment;
        }
        private static ViewItem FindViewItem(ViewItemsCollection viewItems, IModelLayoutViewItem layoutItemModel) {
            IModelViewItem viewItem = layoutItemModel.ViewItem;
            string viewItemId = viewItem != null ? viewItem.Id : layoutItemModel.Id;
            return viewItems[viewItemId];
        }
        private System.Web.UI.WebControls.Unit CalculateCaptionWidth(ViewItem viewItem, ViewItemsCollection viewItems, IModelLayoutViewItem layoutItemModel) {
            var item = layoutItemModel as IModelWebLayoutItem;
            if (item != null) {
                if (item.TextAlignMode == TextAlignModeItem.AutoSize) {
                    return this.GetMaxStringWidth(new string[] { this.EnsureCaptionColon(viewItem.Caption) });
                } else {
                    IModelViewLayoutElement current = layoutItemModel;
                    while (current != null) {
                        var group = current.Parent as IModelWebLayoutGroup;
                        if (group != null) {
                            if (group.TextAlignMode == TextAlignModeGroup.AutoSize) {
                                return this.GetMaxStringWidth(new string[] { this.EnsureCaptionColon(viewItem.Caption) });
                            }
                            if (group.TextAlignMode == TextAlignModeGroup.AlignLocal) {
                                return CalculateLayoutItemCaptionWidthNew((IModelLayoutGroup)group, viewItems, false);
                            }
                            if (group.TextAlignMode == TextAlignModeGroup.AlignWithChildren) {
                                return CalculateLayoutItemCaptionWidthNew((IModelLayoutGroup)group, viewItems, true);
                            }
                        }
                        current = current.Parent as IModelViewLayoutElement;
                    }
                }
            }
            return this.LayoutItemCaptionWidth;
        }
        private System.Web.UI.WebControls.Unit CalculateLayoutItemCaptionWidthNew(IEnumerable<IModelViewLayoutElement> layoutInfo, ViewItemsCollection viewItems, bool recursively) {
            List<string> list = new List<string>();
            CollectLayoutItemVisibleCaptions<IModelViewLayoutElement>(list, layoutInfo, viewItems, recursively);
            return this.GetMaxStringWidth(list);
        }
        private void CollectLayoutItemVisibleCaptions<T>(IList<string> captions, IEnumerable<T> layoutInfo, ViewItemsCollection viewItems, bool recursively) {
            foreach (T itemInfo in layoutInfo) {
                if (itemInfo is IModelLayoutViewItem) {
                    IModelLayoutViewItem layoutItemModel = (IModelLayoutViewItem)itemInfo;
                    ViewItem viewItem = FindViewItem(viewItems, layoutItemModel);
                    if (viewItem != null && GetIsLayoutItemCaptionVisible(layoutItemModel, viewItem) && GetIsItemForCaptionCalculation(layoutItemModel, viewItem)) {
                        MarkRequiredFieldCaptionEventArgs args = new MarkRequiredFieldCaptionEventArgs(viewItem, false);
                        OnMarkRequiredFieldCaption(args);
                        captions.Add(BuildItemCaption(viewItem, args.NeedMarkRequiredField, args.RequiredFieldMark));
                    }
                } else if (recursively) {
                    if (itemInfo is IEnumerable<IModelViewLayoutElement>) {
                        CollectLayoutItemVisibleCaptions<IModelViewLayoutElement>(captions, (IEnumerable<IModelViewLayoutElement>)itemInfo, viewItems, recursively);
                    } else if (itemInfo is IEnumerable<IModelLayoutGroup>) {
                        CollectLayoutItemVisibleCaptions<IModelLayoutGroup>(captions, (IEnumerable<IModelLayoutGroup>)itemInfo, viewItems, recursively);
                    }
                }
            }
        }

    }
}

File: VB/WebLayoutSolution.Module.Web/WebLayoutManagerEx.vb
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Layout
Imports DevExpress.ExpressApp.Model
Imports DevExpress.ExpressApp.Web.Layout
Imports DevExpress.ExpressApp.Web.SystemModule
Imports DevExpress.Persistent.Base
Imports System.Collections.Generic
Imports System.ComponentModel

'#pragma warning disable 109
Namespace WebLayoutSolution.Module.Web

    Public Enum TextAlignModeGroup
        UseParentOptions = 0
        AlignLocal = 1
        AutoSize = 2
        'CustomSize = 3,
        AlignWithChildren = 4
    End Enum

    Public Enum TextAlignModeItem
        UseParentOptions = 0
        'AlignLocal = 1,
        AutoSize = 2
    'CustomSize = 3,
    'AlignWithChildren = 4
    End Enum

    Public Interface IModelWebLayoutGroup

        <Category("Behavior"), DefaultValue(TextAlignModeGroup.UseParentOptions)>
        Property TextAlignMode As TextAlignModeGroup

    End Interface

    Public Interface IModelWebLayoutItem

        <Category("Behavior"), DefaultValue(TextAlignModeItem.UseParentOptions)>
        Property TextAlignMode As TextAlignModeItem

    End Interface

    Public Class WebLayoutManagerEx
        Inherits WebLayoutManager

        Public Sub New(ByVal simple As Boolean, ByVal delayedItemsInitialization As Boolean)
            MyBase.New(simple, delayedItemsInitialization)
        End Sub

        Protected Overrides Function LayoutItem(ByVal viewItems As ViewItemsCollection, ByVal layoutItemModel As IModelLayoutViewItem) As LayoutItemTemplateContainer
            Dim templateContainer As LayoutItemTemplateContainer = New LayoutItemTemplateContainer(Me, viewItems, layoutItemModel)
            templateContainer.Template = LayoutItemTemplate
            templateContainer.ID = WebIdHelper.GetCorrectedLayoutItemId(layoutItemModel)
            Dim viewItem As ViewItem = FindViewItem(viewItems, layoutItemModel)
            templateContainer.ViewItem = viewItem
            If viewItem IsNot Nothing Then
                Dim args As MarkRequiredFieldCaptionEventArgs = New MarkRequiredFieldCaptionEventArgs(viewItem, False)
                OnMarkRequiredFieldCaption(args)
                templateContainer.Caption = BuildItemCaption(viewItem, args.NeedMarkRequiredField, args.RequiredFieldMark)
            End If

            templateContainer.ShowCaption = GetIsLayoutItemCaptionVisible(layoutItemModel, viewItem)
            templateContainer.CaptionWidth = CalculateCaptionWidth(viewItem, viewItems, layoutItemModel)
            templateContainer.CaptionLocation = GetCaptionLocation(layoutItemModel)
            templateContainer.CaptionHorizontalAlignment = GetCaptionHorizontalAlignment(layoutItemModel)
            templateContainer.CaptionVerticalAlignment = GetCaptionVerticalAlignment(layoutItemModel)
            OnLayoutItemCreatedNew(templateContainer, layoutItemModel, viewItem)
            Return templateContainer
        End Function

        Private Sub OnLayoutItemCreatedNew(ByVal templateContainer As LayoutItemTemplateContainerBase, ByVal layoutItemModel As IModelViewLayoutElement, ByVal viewItem As ViewItem)
            If Not DelayedItemsInitialization Then
                templateContainer.Instantiate()
            End If

            OnLayoutItemCreated(New ItemCreatedEventArgs(layoutItemModel, viewItem, templateContainer))
            OnCustomizeAppearance(New CustomizeAppearanceEventArgs(layoutItemModel.Id, New WebLayoutItemAppearanceAdapter(templateContainer), Nothing))
        End Sub

        Private Function GetCaptionLocation(ByVal layoutItemModel As IModelLayoutViewItem) As DevExpress.Utils.Locations
            Dim captionLocation As Locations = layoutItemModel.CaptionLocation
            Return If(Equals(captionLocation, Locations.Default), DefaultLayoutItemCaptionLocation, CType(captionLocation, DevExpress.Utils.Locations))
        End Function

        Private Function GetCaptionHorizontalAlignment(ByVal layoutItemModel As IModelLayoutViewItem) As DevExpress.Utils.HorzAlignment
            Dim captionHorizontalAlignment As DevExpress.Utils.HorzAlignment = layoutItemModel.CaptionHorizontalAlignment
            Return If(Equals(captionHorizontalAlignment, DevExpress.Utils.HorzAlignment.Default), DefaultLayoutItemCaptionHorizontalAlignment, captionHorizontalAlignment)
        End Function

        Private Function GetCaptionVerticalAlignment(ByVal layoutItemModel As IModelLayoutViewItem) As DevExpress.Utils.VertAlignment
            Dim captionVerticalAlignment As DevExpress.Utils.VertAlignment = layoutItemModel.CaptionVerticalAlignment
            Return If(Equals(captionVerticalAlignment, DevExpress.Utils.VertAlignment.Default), DefaultLayoutItemCaptionVerticalAlignment, captionVerticalAlignment)
        End Function

        Private Shared Function FindViewItem(ByVal viewItems As ViewItemsCollection, ByVal layoutItemModel As IModelLayoutViewItem) As ViewItem
            Dim viewItem As IModelViewItem = layoutItemModel.ViewItem
            Dim viewItemId As String = If(viewItem IsNot Nothing, viewItem.Id, layoutItemModel.Id)
            Return viewItems(viewItemId)
        End Function

        Private Function CalculateCaptionWidth(ByVal viewItem As ViewItem, ByVal viewItems As ViewItemsCollection, ByVal layoutItemModel As IModelLayoutViewItem) As System.Web.UI.WebControls.Unit
            Dim item = TryCast(layoutItemModel, IModelWebLayoutItem)
            If item IsNot Nothing Then
                If item.TextAlignMode = TextAlignModeItem.AutoSize Then
                    Return GetMaxStringWidth(New String() {EnsureCaptionColon(viewItem.Caption)})
                Else
                    Dim current As IModelViewLayoutElement = layoutItemModel
                    While current IsNot Nothing
                        Dim group = TryCast(current.Parent, IModelWebLayoutGroup)
                        If group IsNot Nothing Then
                            If group.TextAlignMode = TextAlignModeGroup.AutoSize Then
                                Return GetMaxStringWidth(New String() {EnsureCaptionColon(viewItem.Caption)})
                            End If

                            If group.TextAlignMode = TextAlignModeGroup.AlignLocal Then
                                Return CalculateLayoutItemCaptionWidthNew(CType(group, IModelLayoutGroup), viewItems, False)
                            End If

                            If group.TextAlignMode = TextAlignModeGroup.AlignWithChildren Then
                                Return CalculateLayoutItemCaptionWidthNew(CType(group, IModelLayoutGroup), viewItems, True)
                            End If
                        End If

                        current = TryCast(current.Parent, IModelViewLayoutElement)
                    End While
                End If
            End If

            Return Me.LayoutItemCaptionWidth
        End Function

        Private Function CalculateLayoutItemCaptionWidthNew(ByVal layoutInfo As IEnumerable(Of IModelViewLayoutElement), ByVal viewItems As ViewItemsCollection, ByVal recursively As Boolean) As System.Web.UI.WebControls.Unit
            Dim list As List(Of String) = New List(Of String)()
            Call CollectLayoutItemVisibleCaptions(Of IModelViewLayoutElement)(list, layoutInfo, viewItems, recursively)
            Return GetMaxStringWidth(list)
        End Function

        Private Overloads Sub CollectLayoutItemVisibleCaptions(Of T)(ByVal captions As IList(Of String), ByVal layoutInfo As IEnumerable(Of T), ByVal viewItems As ViewItemsCollection, ByVal recursively As Boolean)
            For Each itemInfo As T In layoutInfo
                If TypeOf itemInfo Is IModelLayoutViewItem Then
                    Dim layoutItemModel As IModelLayoutViewItem = CType(itemInfo, IModelLayoutViewItem)
                    Dim viewItem As ViewItem = FindViewItem(viewItems, layoutItemModel)
                    If viewItem IsNot Nothing AndAlso GetIsLayoutItemCaptionVisible(layoutItemModel, viewItem) AndAlso GetIsItemForCaptionCalculation(layoutItemModel, viewItem) Then
                        Dim args As MarkRequiredFieldCaptionEventArgs = New MarkRequiredFieldCaptionEventArgs(viewItem, False)
                        OnMarkRequiredFieldCaption(args)
                        captions.Add(BuildItemCaption(viewItem, args.NeedMarkRequiredField, args.RequiredFieldMark))
                    End If
                ElseIf recursively Then
                    If TypeOf itemInfo Is IEnumerable(Of IModelViewLayoutElement) Then
                        Call CollectLayoutItemVisibleCaptions(Of IModelViewLayoutElement)(captions, CType(itemInfo, IEnumerable(Of IModelViewLayoutElement)), viewItems, recursively)
                    ElseIf TypeOf itemInfo Is IEnumerable(Of IModelLayoutGroup) Then
                        Call CollectLayoutItemVisibleCaptions(Of IModelLayoutGroup)(captions, CType(itemInfo, IEnumerable(Of IModelLayoutGroup)), viewItems, recursively)
                    End If
                End If
            Next
        End Sub
    End Class
End Namespace
